### PR TITLE
add stub code to show implement event-dispatcher on swoole worker event

### PR DESCRIPTION
**Pull request by eranoitulover**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

As shown in this [issue #1]([https://github.com/mezzio/mezzio-swoole/issues/1), I'm try to create a draft code. Please give me a hint or feedback if something missing. I can rework the code if break some major convention.

I'm also find there's no definition on **workerstop** or **workererror** event. I'm forgot if this remove on central repo, or this code is added by me, cause in my case, I need define this event to closing that opened on worker start.

### Use Case

```php
class SomeEvent
{
    public function __invoke(ContainerInterface $container)
    {
        $listener = $container->get(WorkerListenerProviderInterface::class);
        $listener->addListener(OnWorkerStartEvent::class, function (OnWorkerStartEvent $event) {
            $queue = \GuzzleHttp\Promise\queue();
            $workerServer = $event->getHttpServer();
            $workerServer->tick(1000, function () use ($queue) {
                $queue->run();
            });
        });
    }
}
```

Also can somebody tell me how to test? I'm always get this error on composer test 
```bash
'Exception Thrown: failed to listen server port[127.0.0.1:8080], Error: Invalid argument[22]'.
```

### CODE REVIEW

**Comment by weierophinney** [comment]
You cannot start a Swoole server within a PHPUnit test suite, plain and simple. As such, you need to mock things like the HTTP server.

BTW, I have a working package that does exactly this, and is fully tested: https://github.com/phly/phly-swoole-taskworker

I haven't brought it to mezzio-swoole yet, as it's actually a _lot_ of code, and it can be used with or without Mezzio itself (i.e., it could be used in a vanilla Swoole application if desired).

File: src/SwooleRequestHandlerRunner.php [hunk 1]
```diff
@@ -213,6 +230,21 @@ public function onWorkerStart(SwooleHttpServer $server, int $workerId) : void
             'cwd' => getcwd(),
             'pid' => $workerId,
         ]);
+
+        if ($this->dispatcher) {
+            $this->dispatcher->dispatch(new OnWorkerStartEvent($server, $workerId));
+        }
```
Comment (weierophinney) [inline src/SwooleRequestHandlerRunner.php:227]: Why is this not done in the `onWorkerStart()` method?

**Review by weierophinney — COMMENTED** [review]
I may have misread this when I first reviewed.

Looking at it again, this is code that:

- Optionally adds a PSR-14 dispatcher to the HTTP handler instance.
- Which, when present, triggers an "on worker start" event as the handler starts handling events, and
- registers "workerstop" and "workererror" listeners with the HTTP server that use the dispatcher to trigger userland listeners.

Is that about correct?

I'm not 100% sure we want to support this. Most of this **can** be done in userland (write a delegator on the Swoole HTTP Server service that registers the additional listeners); the only part that would be problematic is triggering the `OnWorkerStartEvent`. For that latter, we could accommodate that with a patch that accepts a callable to trigger with the server instance and worker ID.

On the flip-side, something that _might_ be interesting is to have each of our internal Swoole HTTP Server listeners use a PSR-14 instance to dispatch events, and then provide a default listener provider out-of-the-box. This would allow users like yourself to then register any listeners you want on any of the various HTTP server events.

Thoughts?

File: src/SwooleRequestHandlerRunner.php (hunk 1 above)
Comment (eranoitulover) [reply src/SwooleRequestHandlerRunner.php:236]: ah my bad, it's already executed, so there's no need to change with onworkerstart event, but there's no mechanism yet to handle workerstop/workererror event right?

Comment (eranoitulover) [reply src/SwooleRequestHandlerRunner.php:227]: I think what i propose is overlapping with an already existing feature like `onWorkerStart()` method in `SwooleHttpServer`, But I'm not try to inject directly `onWorkerStart` event at `SwooleHttpServer `object, instead trigger an event in `SwooleRequestHandlerRunner `object which also follow the `SwooleHttpServer `event.

The `onWorkerStart `event at `SwooleHttpServer ` can't be register multiple event right?

**Comment by eranoitulover** [comment]
I just realize that the confusion is appear because we have same "event" at SwooleHttpServer also at SwooleRequestHandlerRunner. Using WorkerListenerProvider, user can register event outside event at SwooleHttpServer, which I assume is restricted to just register one event.

**Comment by weierophinney** [comment]
I repointed your patch to the 2.7.x branch, and have rebased for you; pull the changes before doing more work! :smile:

> I just realize that the confusion is appear because we have same "event" at SwooleHttpServer also at SwooleRequestHandlerRunner. Using WorkerListenerProvider, user can register event outside event at SwooleHttpServer, which I assume is restricted to just register one event.

That's basically what I gathered towards the end.

Swoole allows exactly one listener per event, which is done primarily to prevent problems that occur when multiple listeners handle the same request, task, etc. However, this is often something you actually want to do, which your patch addresses by having each Swoole listener dispatch an event.

My request is that you expand this past the worker start/stop to the full server lifecycle. This will provide consistency, as users will know that this is how they can change how internals work. It also provides us with the ability to hook in things like logging, which would be fantastic.

**Comment by eranoitulover** [comment]
I'm not sure what do you mean about full server lifecycle. Based on this [docs](https://www.swoole.co.uk/docs/modules/swoole-http-server-doc) there's 16 events. 

This patch only defined 6 events, I don't know what other events that need to add. I focused on worker event, because there's use case I need to handle at worker level.

**Comment by basz** [comment]
Awesome work here. Exactly what I need to start a timer on workerstart.

File: src/Event/WorkerListenerProvider.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    public function getListenersForEvent(object $event): iterable
+    {
+        foreach ($this->listeners as $eventType => $listeners) {
+            if (! $event instanceof $eventType) {
+                continue;
+            }
+            foreach ($listeners as $listener) {
+                yield $listener;
+            }
```
Comment (acelaya) [inline src/Event/WorkerListenerProvider.php:27]:
```suggestion
            yield from $listeners;
```

**Comment by acelaya** [comment]
This looks nice. Is it intended to also support custom "start" and "shutdown" events? (I guess it can be done on a different PR after this one has been merged anyway).

I have found a use case for those. I'm running an e2e test suite that starts the app through a swoole server, runs some tests that mainly perform HTTP requests to the API exposed through an HTTP client, and then tear down the server.

Problem is that it's not easy to collect code coverage, so I was trying to do it by starting a coverage collector "on start", and stopping+dumping the result "on shutdown".

---

Edit: Ok, just saw this comment https://github.com/mezzio/mezzio-swoole/pull/24#issuecomment-689810183, haha

**Comment by eranoitulover** [comment]
All discussion should be targeted to #28

**Pull request closed without merging**